sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,205 @@
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NOISE_RMS)
+
+#if !defined MAG_CALIB_UKF_HOTSTART
+#define MAG_CALIB_UKF_HOTSTART FALSE
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART)
+
+#if !defined MAG_CALIB_UKF_HOTSTART_SAVE_FILE
+#define MAG_CALIB_UKF_HOTSTART_SAVE_FILE /data/ftp/internal_000/mag_ukf_calib.txt
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART_SAVE_FILE)
+
+#if !defined MAG_CALIB_UKF_VERBOSE || !MAG_CALIB_UKF_VERBOSE
+#define VERBOSE_PRINT(...)
+#else
+#define VERBOSE_PRINT PRINT
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_VERBOSE)
+
+bool settings_reset_state = false;
+
+static TRICAL_instance_t mag_calib;
+static abi_event mag_ev;
+static abi_event h_ev;
+
+static uint32_t timestamp_geo_mag;
+static struct FloatVect3 H = { .x = AHRS_H_X, .y = AHRS_H_Y, .z =  AHRS_H_Z};
+
+#if MAG_CALIB_UKF_HOTSTART
+static FILE *fp;
+static char hotstart_file_name[512];
+#endif
+
+void mag_calib_ukf_init(void)
+{
+  TRICAL_init(&mag_calib);
+  TRICAL_norm_set(&mag_calib, MAG_CALIB_UKF_NORM);
+  TRICAL_noise_set(&mag_calib, MAG_CALIB_UKF_NOISE_RMS);
+
+  mag_calib_hotstart_read();
+  AbiBindMsgIMU_MAG_INT32(IMU_BOARD_ID, &mag_ev, mag_calib_ukf_run);
+  AbiBindMsgGEO_MAG(ABI_BROADCAST, &h_ev, mag_calib_update_field);    ///< GEO_MAG_SENDER_ID is defined in geo_mag.c so unknown
+}
+
+void mag_calib_ukf_run(uint8_t sender_id, uint32_t stamp, struct Int32Vect3 *mag)
+{
+  float measurement[3] = {0.0, 0.0, 0.0}, calibrated_measurement[3] = {0.0, 0.0, 0.0};
+  if (sender_id != MAG_CALIB_UKF_ID) {
+    /** Update geo_mag based on MAG_CALIB_UKF_GEO_MAG_TIMEOUT (0 = no periodic updates) **/
+    if (MAG_CALIB_UKF_GEO_MAG_TIMEOUT && GpsFixValid() && (get_sys_time_msec() - timestamp_geo_mag) >= 1000 * MAG_CALIB_UKF_GEO_MAG_TIMEOUT) {
+      geo_mag.ready     = false;
+      geo_mag.calc_once = true;   ///< Geo_mag will not re-update the calculation when the throttle is on so this is neccesary
+    }
+    /** See if we need to reset the state **/
+    if (settings_reset_state) {
+      TRICAL_reset(&mag_calib);
+      settings_reset_state = false;
+    }
+    /** Update magnetometer UKF and calibrate measurement **/
+    if (mag->x != 0 || mag->y != 0 || mag->z != 0) {
+      /** Rotate the local magnetic field by our current attitude **/
+      struct FloatQuat *body_quat = stateGetNedToBodyQuat_f();
+      struct FloatVect3 expected_measurement;
+      float_quat_vmult(&expected_measurement, body_quat, &H);
+      float expected_mag_field[3] = { expected_measurement.x, expected_measurement.y, expected_measurement.z };
+      /** Update magnetometer UKF **/
+      measurement[0] = MAG_FLOAT_OF_BFP(mag->x);
+      measurement[1] = MAG_FLOAT_OF_BFP(mag->y);
+      measurement[2] = MAG_FLOAT_OF_BFP(mag->z);
+      TRICAL_estimate_update(&mag_calib, measurement, expected_mag_field);
+      TRICAL_measurement_calibrate(&mag_calib, measurement, calibrated_measurement);
+      /** Save calibrated result **/
+      mag->x = (int32_t) MAG_BFP_OF_REAL(calibrated_measurement[0]);
```

Comment:
That would be ok for now (as long as we still have the global struct).
For the copy simply make a global struct for the mag values.